src/main/scala/CustomBootPin.scala

```diff
@@ -0,0 +1,87 @@
+package testchipip
+
+import chisel3._
+import chisel3.util._
+import chisel3.experimental.{IO}
+import freechips.rocketchip.config._
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.devices.tilelink._
+import freechips.rocketchip.regmapper._
+import freechips.rocketchip.subsystem._
+import freechips.rocketchip.util._
+import freechips.rocketchip.tile._
+import freechips.rocketchip.prci._
+
+case class CustomBootPinParams(
+  customBootAddress: BigInt = 0x80000000L, // Default is DRAM_BASE
+  masterWhere: TLBusWrapperLocation = PBUS // This needs to write to clint and bootaddrreg, which are probably on PBUS
+)
+
+case object CustomBootPinKey extends Field[Option[CustomBootPinParams]](Some(CustomBootPinParams()))
+
+class WithCustomBootPinAltAddr(address: BigInt = 0x10040) extends Config((site, here, up) => {
+  case CustomBootPinKey => up(CustomBootPinKey, site).map(p => p.copy(customBootAddress = address))
+})
+
+trait CanHavePeripheryCustomBootPin { this: BaseSubsystem =>
+  val custom_boot_pin = p(CustomBootPinKey).map { params =>
+    val tlbus = locateTLBusWrapper(params.masterWhere)
+    val clientParams = TLMasterPortParameters.v1(
+      clients = Seq(TLMasterParameters.v1(
+        name = "custom-boot",
+        sourceId = IdRange(0, 1)
```

Comment:
I believe your state machine requires `minLatency = 1` as it can't handle combo A->D